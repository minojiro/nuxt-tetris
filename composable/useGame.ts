import { ref, computed } from "vue";
import { BLOCKS } from "./blocks";

export type Field = number[][];

export type Direction = "rotate" | "down" | "left" | "right";

export type GameStatus = "cover" | "playing" | "end";

type Position = [x: number, y: number];

const KEY_MOVE_MAP: Record<Direction, Position> = {
  rotate: [0, 0],
  down: [0, 1],
  left: [-1, 0],
  right: [1, 0],
};

const FIELD_WIDTH = 12;
const FIELD_HEIGHT = 22;

const getEmptyField = (withWall = false): Field => {
  return Array.from({ length: FIELD_HEIGHT }, (_, i) =>
    Array.from({ length: FIELD_WIDTH }, (_, j) => {
      return withWall &&
        (i === FIELD_HEIGHT - 1 || j === 0 || j === FIELD_WIDTH - 1)
        ? 1
        : 0;
    })
  );
};

export const useGame = ({ genRandomNum }: { genRandomNum: () => number }) => {
  const gameStatus = ref<GameStatus>("cover");

  const fixedField = ref(getEmptyField(true));
  const currentBlockPosition = ref<Position>([0, 0]);
  const currentBlockId = ref<number>(0);
  const currentBlockRotate = ref<number>(0);

  const fieldAndConflict = computed(() => {
    let conflict = false;
    if (gameStatus.value !== "playing") {
      const isCover = gameStatus.value === "cover";
      const field = isCover ? getEmptyField(true) : fixedField.value;
      return {
        field,
        conflict,
      };
    }
    const [blockX, blockY] = currentBlockPosition.value;
    const block =
      BLOCKS[currentBlockId.value][
        currentBlockRotate.value % BLOCKS[currentBlockId.value].length
      ];
    const field = fixedField.value.map((line, y) =>
      line.map((fixedFieldCell, x) => {
        const blockCell = block[y - blockY]?.[x - blockX] || 0;
        conflict = conflict || !!(blockCell && fixedFieldCell);
        return blockCell + fixedFieldCell;
      })
    );
    return { field, conflict };
  });

  const field = computed(() => fieldAndConflict.value.field);
  const conflict = computed(() => fieldAndConflict.value.conflict);

  const setNextBlock = () => {
    currentBlockPosition.value = [4, 1];
    currentBlockId.value = Math.floor(genRandomNum() * BLOCKS.length);
    currentBlockRotate.value = 0;

    if (conflict.value) {
      gameStatus.value = "end";
    }
  };

  const moveBlock = (direction: Direction) => {
    if (gameStatus.value !== "playing") return;
    const [movX, movY] = KEY_MOVE_MAP[direction];
    const [curX, curY] = currentBlockPosition.value;
    currentBlockPosition.value = [curX + movX, curY + movY];
    if (direction === "rotate") {
      currentBlockRotate.value++;
    }

    if (conflict.value) {
      currentBlockPosition.value = [curX, curY];
      if (direction === "down") {
        fixedField.value = field.value;
        setNextBlock();
      } else if (direction === "rotate") {
        currentBlockRotate.value--;
      }
    }
  };

  const gameStart = () => {
    gameStatus.value = "playing";
    setNextBlock();
  };

  const tick = () => {
    moveBlock("down");
  };

  return {
    field,
    gameStatus: computed(() => gameStatus.value),
    moveBlock,
    gameStart,
    tick,
  };
};
