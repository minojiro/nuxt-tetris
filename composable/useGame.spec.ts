import { describe, it, expect } from "vitest";
import { useGame, Direction } from "./useGame";
import { BLOCKS } from "./blocks";

const getFieldStr = (field: number[][]) => {
  return field
    .map((line, i) => line.map((n) => (n ? n : " ")).join("") + i)
    .join("\n");
};

describe("useGame", () => {
  it("フィールドの配列を取得できること", () => {
    const genRandomNum = () => 0;
    const { field } = useGame({ genRandomNum });
    expect(field.value).toHaveLength(22);
    expect(field.value[0]).toHaveLength(12);
  });

  it("ゲーム開始前は moveBlock しても変化がないこと", () => {
    const genRandomNum = () => 0;
    const getRandomDir = (): Direction => {
      const dirs: Direction[] = ["left", "right", "down", "rotate"];
      return dirs[~~(Math.random() * dirs.length)];
    };
    const { field, moveBlock } = useGame({ genRandomNum });
    const fieldBefore = getFieldStr(field.value);
    for (let i = 0; i < 100; i++) {
      const dir = getRandomDir();
      moveBlock(dir);
    }
    const fieldAfter = getFieldStr(field.value);
    expect(fieldAfter).toBe(fieldBefore);
  });

  describe("ブロックの操作", () => {
    const genRandomNum = () => 0;

    it.each(["left", "right"])(
      "%s に移動できて、壁は突き抜けないこと",
      (dir) => {
        const { field, moveBlock, gameStart } = useGame({ genRandomNum });
        gameStart();
        for (var i = 1; i <= 5; i++) {
          moveBlock(dir as Direction);
          if (i > 2) {
            expect(getFieldStr(field.value)).toMatchSnapshot();
          }
        }
      }
    );

    it("下に移動できて、一番下まで行ったら次のブロックが追加されること", () => {
      const { field, moveBlock, gameStart } = useGame({ genRandomNum });
      gameStart();
      for (var i = 1; i <= 18; i++) {
        moveBlock("down");
        if (i > 15) {
          expect(getFieldStr(field.value)).toMatchSnapshot();
        }
      }
    });

    it("回転できること", () => {
      const genRandomNum = () => (1 / BLOCKS.length) * 4;
      const { field, moveBlock, gameStart } = useGame({ genRandomNum });
      gameStart();
      for (var i = 1; i <= 5; i++) {
        moveBlock("rotate");
        expect(getFieldStr(field.value)).toMatchSnapshot();
      }
    });
  });

  it("新しいブロックが置けないと終了すること", () => {
    const genRandomNum = () => 0;
    const { tick, field, gameStart, gameStatus } = useGame({
      genRandomNum,
    });
    expect(gameStatus.value).toBe("cover");
    gameStart();
    expect(gameStatus.value).toBe("playing");
    for (var i = 1; i <= 100; i++) {
      tick();
    }
    expect(getFieldStr(field.value)).toMatchSnapshot();
    expect(gameStatus.value).toBe("end");
  });
});
