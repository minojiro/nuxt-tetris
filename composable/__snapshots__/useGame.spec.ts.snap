// Vitest Snapshot v1

exports[`useGame > ブロックの操作 > left に移動できて、壁は突き抜けないこと 1`] = `
"1111111111110
1          11
1 11       12
1 11       13
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1          119
1          120
11111111111121"
`;

exports[`useGame > ブロックの操作 > left に移動できて、壁は突き抜けないこと 2`] = `
"1111111111110
1          11
111        12
111        13
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1          119
1          120
11111111111121"
`;

exports[`useGame > ブロックの操作 > left に移動できて、壁は突き抜けないこと 3`] = `
"1111111111110
1          11
111        12
111        13
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1          119
1          120
11111111111121"
`;

exports[`useGame > ブロックの操作 > right に移動できて、壁は突き抜けないこと 1`] = `
"1111111111110
1          11
1       11 12
1       11 13
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1          119
1          120
11111111111121"
`;

exports[`useGame > ブロックの操作 > right に移動できて、壁は突き抜けないこと 2`] = `
"1111111111110
1          11
1        1112
1        1113
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1          119
1          120
11111111111121"
`;

exports[`useGame > ブロックの操作 > right に移動できて、壁は突き抜けないこと 3`] = `
"1111111111110
1          11
1        1112
1        1113
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1          119
1          120
11111111111121"
`;

exports[`useGame > ブロックの操作 > 下に移動できて、一番下まで行ったら次のブロックが追加されること 1`] = `
"1111111111110
1          11
1          12
1          13
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1    11    118
1    11    119
1          120
11111111111121"
`;

exports[`useGame > ブロックの操作 > 下に移動できて、一番下まで行ったら次のブロックが追加されること 2`] = `
"1111111111110
1          11
1          12
1          13
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1    11    119
1    11    120
11111111111121"
`;

exports[`useGame > ブロックの操作 > 下に移動できて、一番下まで行ったら次のブロックが追加されること 3`] = `
"1111111111110
1          11
1    11    12
1    11    13
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1    11    119
1    11    120
11111111111121"
`;

exports[`useGame > ブロックの操作 > 回転できること 1`] = `
"1111111111110
1    55    11
1    5     12
1    5     13
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1          119
1          120
11111111111121"
`;

exports[`useGame > ブロックの操作 > 回転できること 2`] = `
"1111111111110
1          11
1    555   12
1      5   13
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1          119
1          120
11111111111121"
`;

exports[`useGame > ブロックの操作 > 回転できること 3`] = `
"1111111111110
1     5    11
1     5    12
1    55    13
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1          119
1          120
11111111111121"
`;

exports[`useGame > ブロックの操作 > 回転できること 4`] = `
"1111111111110
1          11
1    5     12
1    555   13
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1          119
1          120
11111111111121"
`;

exports[`useGame > ブロックの操作 > 回転できること 5`] = `
"1111111111110
1    55    11
1    5     12
1    5     13
1          14
1          15
1          16
1          17
1          18
1          19
1          110
1          111
1          112
1          113
1          114
1          115
1          116
1          117
1          118
1          119
1          120
11111111111121"
`;

exports[`useGame > 新しいブロックが置けないと終了すること 1`] = `
"1111111111110
1          11
1          12
1    11    13
1    11    14
1    11    15
1    11    16
1    11    17
1    11    18
1    11    19
1    11    110
1    11    111
1    11    112
1    11    113
1    11    114
1    11    115
1    11    116
1    11    117
1    11    118
1    11    119
1    11    120
11111111111121"
`;
